### PR TITLE
Log information about received messages

### PR DESCRIPTION
**Pull request by jrusbatch**
Necessary for calculating participation statistics.

#150

### CODE REVIEW

File: Modix.Data/Models/Core/MessageEntity.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace Modix.Data.Models.Core
+{
+    public class MessageEntity
+    {
+        [Key]
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public long Id { get; set; }
```
Comment (JakenVeina) [inline Modix.Data/Models/Core/MessageEntity.cs]: No surrogate keys, please. `MessageId` should be the primary key, as it uniquely identifies every single message, anywhere.

Have a look at the other tracking tables for reference, if you want.

File: Modix.Data/Models/Core/MessageEntity.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        public ulong MessageId { get; set; }
+
+        [Required]
+        public ulong GuildId { get; set; }
+
+        [Required]
+        public ulong ChannelId { get; set; }
+
+        [Required]
+        public ulong UserId { get; set; }
```
Comment (JakenVeina) [inline Modix.Data/Models/Core/MessageEntity.cs]: This should be `AuthorId`, for clarity.

Comment (JakenVeina) [reply Modix.Data/Models/Core/MessageEntity.cs]: Also, you should have a navigation property `Author`

File: Modix.Data/Models/Core/MessageEntity.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        public long Id { get; set; }
+
+        [Required]
+        public ulong MessageId { get; set; }
+
+        [Required]
+        public ulong GuildId { get; set; }
+
+        [Required]
+        public ulong ChannelId { get; set; }
```
Comment (JakenVeina) [inline Modix.Data/Models/Core/MessageEntity.cs]: You should add a navigation property `Channel`.

File: Modix.Data/Models/Core/MessageEntity.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        [Required]
+        public ulong UserId { get; set; }
+
+        [Required]
+        public DateTimeOffset Timestamp { get; set; }
+
+        [Required]
+        public string OriginalMessageHash { get; set; }
```
Comment (JakenVeina) [inline Modix.Data/Models/Core/MessageEntity.cs]: Is string the most appropriate type for the hash? Is there a hashing algorithm we could use that spits out a `long`?

File: Modix.Data/ModixContext.cs [hunk 5]
```diff
@@ -127,6 +129,13 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
                 .Entity<DesignatedChannelMappingEntity>()
                 .Property(x => x.ChannelDesignation)
                 .HasConversion(new EnumToStringConverter<ChannelDesignation>());
+
+            modelBuilder.Entity<MessageEntity>()
+                .HasIndex(x => x.MessageId)
+                .IsUnique();
```
Comment (JakenVeina) [inline Modix.Data/ModixContext.cs]: Remove this after changing the primary key, as it'll be redundant.

File: Modix.Data/ModixContext.cs [hunk 6]
```diff
@@ -127,6 +129,13 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
+
+            modelBuilder.Entity<MessageEntity>()
+                .HasIndex(x => new { x.GuildId, x.UserId });
```
Comment (JakenVeina) [inline Modix.Data/ModixContext.cs]: Remove this after adding an `Author` navigation property, as it'll be redundant.

File: Modix.Services/Core/MessageLogBehavior.cs [hunk 7]
```diff
@@ -1,29 +1,36 @@
...
     public class MessageLogBehavior : BehaviorBase
     {
         private readonly DiscordSocketClient _discordClient;
 
-        public MessageLogBehavior(DiscordSocketClient discordClient, IServiceProvider serviceProvider) : base(serviceProvider)
+        public MessageLogBehavior(DiscordSocketClient discordClient, IServiceProvider serviceProvider, ILogger<MessageLogBehavior> logger) : base(serviceProvider)
         {
             _discordClient = discordClient;
+
+            Log = logger ?? NullLogger<MessageLogBehavior>.Instance;
```
Comment (JakenVeina) [inline Modix.Services/Core/MessageLogBehavior.cs]: I thought we used Serilog for logging, which utilizes a `static` class for its API?

Does Serilog support the .NET Core Logging infrastructure? Are you just changing from static coupling to dependency injection? If so, this really belongs in a separate PR.

File: Modix.Services/Core/MessageLogBehavior.cs [hunk 8]
```diff
@@ -54,6 +62,44 @@ private async Task<bool> ShouldSkip(IGuild guild, IMessageChannel channel)
             return result;
         }
 
+        private async Task HandleMessageReceived(SocketMessage message)
+        {
+            Log.LogDebug("Handling message received event for message #{MessageId}.", message.Id);
+
+            if (!message.Content.StartsWith('!') &&
```
Comment (JakenVeina) [inline Modix.Services/Core/MessageLogBehavior.cs]: Please don't hard-code this. Unless there's a legit way to retrieve the configured command prefix within this scope, just log it like any other message.

File: Modix.Services/Core/MessageLogBehavior.cs [hunk 9]
```diff
@@ -54,6 +62,44 @@ private async Task<bool> ShouldSkip(IGuild guild, IMessageChannel channel)
...
+                message.Channel is IGuildChannel channel &&
+                message.Author is IGuildUser author &&
+                !author.IsBot && !author.IsWebhook)
```
Comment (JakenVeina) [inline Modix.Services/Core/MessageLogBehavior.cs]: checking against `IGuildUser` and `IGuildChannel` are 100% not sufficient to check if there is actually a guild value. The only surefire way is to do the full cast and check whether `Guild` is `null`.

File: Modix.Services/Core/MessageLogBehavior.cs [hunk 10]
```diff
@@ -54,6 +62,44 @@ private async Task<bool> ShouldSkip(IGuild guild, IMessageChannel channel)
...
+            {
+                await SelfExecuteRequest<ModixContext>(async db =>
```
Comment (JakenVeina) [inline Modix.Services/Core/MessageLogBehavior.cs]: There should be no interaction with `ModixContext` outside of `Modix.Data`. You need to write a repository interface and class.

File: Modix/Modules/UserInfoModule.cs [hunk 11]
```diff
@@ -7,18 +7,22 @@
...
+using Modix.Data;
 
 namespace Modix.Modules
 {
     public class UserInfoModule : ModuleBase
     {
-        public UserInfoModule(ILogger<UserInfoModule> logger)
+        public UserInfoModule(ModixContext modixContext, ILogger<UserInfoModule> logger)
         {
+            Db = modixContext;
```
Comment (JakenVeina) [inline Modix/Modules/UserInfoModule.cs]: As mentioned above, you need a repository here, not `ModixContext` directly.

File: Modix.Data/Models/Core/MessageEntity.cs (hunk 2 above)
Comment (jrusbatch) [reply Modix.Data/Models/Core/MessageEntity.cs]: I considered a relationship between this table and `GuildUsers`, but decided to remove it because it didn't look like it was safe to assume a guild user would exist before the insert, and because I didn't see a value add from it. Also, I'm not sure it's safe if IDs in other tables are stored as `bigint` instead of `numeric(20, 0)`.

File: Modix.Data/Models/Core/MessageEntity.cs (hunk 4 above)
Comment (jrusbatch) [reply Modix.Data/Models/Core/MessageEntity.cs]: This seemed like the most straightforward approach. What advantage would a long offer over a string?

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 7 above)
Comment (jrusbatch) [reply Modix.Services/Core/MessageLogBehavior.cs]: Yeah, the injected logger logs to Serilog. [`UserInfoModule`](https://github.com/discord-csharp/MODiX/blob/6489654dd0f02fdd474b6c10ac5da5a54dc0515b/Modix/Modules/UserInfoModule.cs#L17) does the same.

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 8 above)
Comment (jrusbatch) [reply Modix.Services/Core/MessageLogBehavior.cs]: Is there a way?

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 10 above)
Comment (jrusbatch) [reply Modix.Services/Core/MessageLogBehavior.cs]: I started to, but it introduced a lot of additional ceremony without a clear benefit. I opted for the simplest approach.

File: Modix/Modules/UserInfoModule.cs (hunk 11 above)
Comment (jrusbatch) [reply Modix/Modules/UserInfoModule.cs]: Responded to the comment above.

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 9 above)
Comment (jrusbatch) [reply Modix.Services/Core/MessageLogBehavior.cs]: I don't actually access a `Guild` property. I use `IGuildChannel.GuildId` which is of type `ulong` (not `ulong?`).

File: Modix.Data/Models/Core/MessageEntity.cs (hunk 2 above)
Comment (JakenVeina) [reply Modix.Data/Models/Core/MessageEntity.cs]: The point about foreign keys is almost certainly valid. I'll add a note to the issue about adding the relationship then.

The user SHOULD exist simply by the fact that we track users on a per-message basis already, but we can definitely avoid the risk of the user not being known. I'd say we simply have `UserTrackingBehavior` not listen to `MessageReceived` and have only `MessageLogBehavior` do that, and call `TrackUserAsync` for the message author, before logging the message.

The advantage to having the relationship is that we can pull back detailed information about a message all in one query, for use in UI's. I.E. if you wanna display the name of the author, you don't have to a separate call to get the name of the user from its id, which would either guarantee another DB hit if we pull from the DB, or risk the user's info not being found if we query from Discord.NET.

On a more fundamental note, the `GuildUsers` and `GuildChannels` tables are specifically intended to track info about all users and channels within any guild the bot serves. If there's a situation where the user or channel is NOT present there, I think that represents a separate bug.

File: Modix.Data/Models/Core/MessageEntity.cs (hunk 4 above)
Comment (JakenVeina) [reply Modix.Data/Models/Core/MessageEntity.cs]: Aside from being smaller, not much I guess. It just struck me as slightly odd. Maybe change the name to make it clear that that's a base64 encoding? Or make it a `byte[]`?

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 10 above)
Comment (JakenVeina) [reply Modix.Services/Core/MessageLogBehavior.cs]: It's indeed a lot of tedium, but I feel it needs to be there because:

A) It's the clearly established pattern.
B) Dependencies against a `DbContext` object are EXTREMELY difficult to decouple. When we start adding testing, anything with a `DbContext` dependency is going to be almost impossible to write tests for. Any tests we DO write are probably going to involve building a fake database of some kind, so they'll still be convoluted.
C) The difference between a good EF query and a terrible EF query are very subtle (regarding how they're translated to SQL), so it's extremely beneficial to keep tight control over where and how EF queries are being written. Having individual queries be well defined as a single method will also make it much easier to test, and even automatically find queries that aren't being translated properly (I.E. we can actually check how many SQL queries were executed).
D) All of the above was discussed at length before any of it was committed into Modix.

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 9 above)
Comment (jrusbatch) [reply Modix.Services/Core/MessageLogBehavior.cs]: Good to know. I'll fix it.

File: Modix.Services/Core/MessageLogBehavior.cs (hunk 8 above)
Comment (jrusbatch) [reply Modix.Services/Core/MessageLogBehavior.cs]: I agree this should probably be a configurable thing, especially if we're already thinking about other guilds would use MODiX. If/when this gets promoted to a setting we should remove this.

Comment (JakenVeina) [reply Modix.Services/Core/MessageLogBehavior.cs]: If you really wanna leave it in, put a TODO on it.
